flink-cdc-e2e-tests/src/test/java/com/ververica/cdc/connectors/e2e/utils/JdbcProxy.java

```diff
@@ -0,0 +1,92 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ververica.cdc.connectors.e2e.utils;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+/** Proxy to communicate with database using JDBC protocal. */
+public class JdbcProxy {
+
+    private final String url;
+    private final String userName;
+    private final String password;
+    private final String driverClass;
+
+    public JdbcProxy(String url, String userName, String password, String driverClass) {
+        this.url = url;
+        this.userName = userName;
+        this.password = password;
+        this.driverClass = driverClass;
+    }
+
+    public void checkResult(List<String> expectedResult, String table, String[] fields)
+            throws SQLException, ClassNotFoundException {
+        Class.forName(driverClass);
+        try (Connection dbConn = DriverManager.getConnection(url, userName, password);
+                PreparedStatement statement = dbConn.prepareStatement("select * from " + table);
+                ResultSet resultSet = statement.executeQuery()) {
+            List<String> results = new ArrayList<>();
+            while (resultSet.next()) {
+                List<String> result = new ArrayList<>();
+                for (String field : fields) {
+                    Object value = resultSet.getObject(field);
+                    if (value == null) {
+                        result.add("null");
+                    } else {
+                        result.add(value.toString());
+                    }
+                }
+
+                results.add(StringUtils.join(result, ","));
+            }
+            Collections.sort(results);
+            Collections.sort(expectedResult);
+            assertEquals(expectedResult, results);
+        }
+    }
+
+    public void checkResultWithTimeout(
+            List<String> expectedResult, String table, String[] fields, long timeout)
+            throws Exception {
+        long endTimeout = System.currentTimeMillis() + timeout;
+        boolean result = false;
+        while (System.currentTimeMillis() < endTimeout) {
+            try {
+                checkResult(expectedResult, table, fields);
+                result = true;
+                break;
+            } catch (AssertionError | SQLException throwable) {
+                Thread.sleep(1000L);
+            }
+        }
+        if (!result) {
+            checkResult(expectedResult, table, fields);
+        }
```

Comment:
A check with timeout exception is better than a final check

Reply:
I don't think so. A check here can print the mimatched results which is much more meaningful than timeout exception.

Reply:
I agree with you, I noticed the timeout check logic may problem, if the first attempt we didn't pull enough data(i.e the upstream program is still writing), the check will fail due to we call `checkResult` at first time in and the timeout is useless?